pkg/inventory/inventory-client.go

```diff
@@ -24,455 +22,287 @@ import (
-		return nil, fmt.Errorf("inventoryInfo must be specified")
-	}
-
-	var clusterInvObjects object.UnstructuredSet
-	var err error
-	switch inv.Strategy() {
-	case NameStrategy:
-		clusterInvObjects, err = cic.getClusterInventoryObjsByName(ctx, inv)
-	case LabelStrategy:
-		clusterInvObjects, err = cic.getClusterInventoryObjsByLabel(ctx, inv)
-	default:
-		panic(fmt.Errorf("unknown inventory strategy: %s", inv.Strategy()))
-	}
-	return clusterInvObjects, err
+// CreateOrUpdate the in-cluster inventory
+// Updates the unstructured object, or creates it if it doesn't exist
+func (cic *UnstructuredClient) CreateOrUpdate(ctx context.Context, inv Inventory, _ UpdateOptions) error {
+	ui, ok := inv.(*UnstructuredInventory)
+	if !ok {
+		return fmt.Errorf("expected UnstructuredInventory")
+	}
+	if ui == nil {
+		return fmt.Errorf("inventory is nil")
+	}
+	return retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		create := false
+		obj, err := cic.client.Namespace(ui.Namespace()).Get(ctx, ui.Name(), metav1.GetOptions{})
+		if apierrors.IsNotFound(err) {
+			create = true
+		} else if err != nil {
+			return err
+		}
+		if obj != nil { // Obj is nil when IsNotFound, in this case keep initial/empty obj
+			ui.ClusterObj = obj
+		}
+		uObj, err := cic.toUnstructured(ui)
+		if err != nil {
+			return err
+		}
+		var newObj *unstructured.Unstructured
+		if create {
+			newObj, err = cic.client.Namespace(uObj.GetNamespace()).Create(ctx, uObj, metav1.CreateOptions{})
+			if err != nil {
+				return err
+			}
+		} else {
+			newObj, err = cic.client.Namespace(uObj.GetNamespace()).Update(ctx, uObj, metav1.UpdateOptions{})
+			if err != nil {
+				return err
+			}
+		}
+		ui.ClusterObj = newObj
+		return nil
+	})
 }
 
-func (cic *ClusterClient) ListClusterInventoryObjs(ctx context.Context) (map[string]object.ObjMetadataSet, error) {
-	// Define the mapping
-	mapping, err := cic.mapper.RESTMapping(cic.gvk.GroupKind(), cic.gvk.Version)
-	if err != nil {
-		return nil, err
+// UpdateStatus updates the status of the in-cluster inventory
+// Performs a simple in-place update on the unstructured object
+func (cic *UnstructuredClient) UpdateStatus(ctx context.Context, inv Inventory, _ UpdateOptions) error {
+	ui, ok := inv.(*UnstructuredInventory)
+	if !ok {
+		return fmt.Errorf("expected UnstructuredInventory")
 	}
-
-	// retrieve the list from the cluster
-	clusterInvs, err := cic.dc.Resource(mapping.Resource).List(ctx, metav1.ListOptions{})
-	if err != nil && !apierrors.IsNotFound(err) {
-		return nil, err
+	if ui == nil {
+		return fmt.Errorf("inventory is nil")
 	}
-	if apierrors.IsNotFound(err) {
-		return map[string]object.ObjMetadataSet{}, nil
-	}
-
-	identifiers := make(map[string]object.ObjMetadataSet)
-
-	for i, inv := range clusterInvs.Items {
-		invName := inv.GetName()
-		identifiers[invName] = object.ObjMetadataSet{}
-		invObj, err := cic.InventoryFactoryFunc(&clusterInvs.Items[i])
+	return retry.RetryOnConflict(retry.DefaultRetry, func() error {
```

Comment:
Since we've used this wrong in the past, perhaps add a comment here to explain that this only makes up to 4 attempts with minimal delay, to avoid needing to retry the whole Apply/Destroy when there's a temporary conflict.

Reply:
Done